README.md
# picamraw
Library for extracting raw bayer data from a Raspberry Pi JPEG+RAW file.

Installable from [PyPI](https://pypi.org/project/picamraw/); usable without camera hardware present.


# Usage example
## Extract raw bayer array
```python
from picamraw import PiRawBayer, PiCameraVersion

raw_bayer = PiRawBayer(
    filepath='path/to/image.jpeg',  # A JPEG+RAW file, e.g. an image captured using raspistill with the "--raw" flag
    camera_version=PiCameraVersion.V2,
    sensor_mode=0
)
raw_bayer.bayer_array   # A 16-bit 2D numpy array of the bayer data
raw_bayer.bayer_order   # A `BayerOrder` enum that describes the arrangement of the R,G,G,B pixels in the bayer_array
raw_bayer.to_rgb()      # A 16-bit 3D numpy array of bayer data collapsed into RGB channels (see docstring for details).
raw_bayer.to_3d()       # A 16-bit 3D numpy array of bayer data split into RGB channels (see docstring for details).
```


# Testing

This package is tested using [`tox`](https://tox.readthedocs.io/).
To run tests, simply `pip install tox` and then run `tox`.

Note: this code has only been tested against an image captured with camera version V2 and sensor_mode 0.


# Attribution
This library was forked from the [PiCamera](https://github.com/waveform80/picamera) package and heavily modified.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org).


## [1.2.0] - 2018-12-19
### Changed
- Fixed bug in assignment of low bits. Pixel value outputs may change by up to 4 DN (4/1024). Thanks to @6by9 for the [report](https://github.com/illes/raspiraw/issues/3).


## [1.1.0] - 2018-12-03
### Added
- `PiRawBayer.to_rgb()` method that collapses the extracted raw bayer 2D array to a 3D RGB array, with no demosaicing other than averaging the two green channels.


## [1.0.0] - 2018-11-29
### Added
- `PiRawBayer` class that extracts raw bayer data out of a JPEG+RAW file.
- `PiRawBayer.to_3d()` method that splits the extracted raw bayer 2D array into a 3D RGB array - moving each pixel into either the R, G, or B channel according to the `PiRawBayer.bayer_order` attribute.

picamraw/main_test.py
import pkg_resources
from unittest.mock import sentinel, MagicMock

import numpy as np
import pytest

from .constants import BayerOrder, PiCameraVersion
from . import main as module

picamv2_jpeg_path = pkg_resources.resource_filename(__name__, 'test_fixtures/picamv2.jpeg')
picamv2_BGGR_bayer_array_path = pkg_resources.resource_filename(__name__, 'test_fixtures/picamv2_BGGR_bayer_array.npy')
picamv2_3d_path = pkg_resources.resource_filename(__name__, 'test_fixtures/picamv2_3d.npy')
picamv2_rgb_path = pkg_resources.resource_filename(__name__, 'test_fixtures/picamv2_rgb.npy')


class TestGetRawBayerBytes:
    def test_get_raw_bayer_bytes(self, mocker):
        mock_stream_with_correct_prefix = b'initial-file-contents-BRCM-test-stream'

        mocker.patch.object(module, '_get_raw_block_size').return_value = 16

        actual = module._get_raw_bayer_bytes(
            jpeg_data_as_bytes=mock_stream_with_correct_prefix,
            camera_version=sentinel.camera_version,
            sensor_mode=sentinel.sensor_mode,
        )

        assert actual == b'BRCM-test-stream'

    def test_get_raw_bayer_bytes__raises_if_missing_prefix(self, mocker):
        with pytest.raises(ValueError):
            mock_stream_with_missing_prefix = b'initial-file-contents-test-stream'

            mocker.patch.object(module, '_get_raw_block_size').return_value = 16
            module._get_raw_bayer_bytes(
                jpeg_data_as_bytes=mock_stream_with_missing_prefix,
                camera_version=sentinel.camera_version,
                sensor_mode=sentinel.sensor_mode,
            )


class TestGetRawBlockSize:
    @pytest.mark.parametrize('camera_version,sensor_mode', [
        (camera_version_enum, sensor_mode)
        for camera_version_enum in PiCameraVersion
        for sensor_mode in range(0, 8)
    ])
    def test_get_raw_block_size__has_values_for_all_camera_versions_and_sensor_modes(self, camera_version, sensor_mode):
        actual = module._get_raw_block_size(camera_version, sensor_mode)
        assert actual is not None


class TestExtractRawFromJpeg:
    bayer_array, bayer_order = module.extract_raw_from_jpeg(
        filepath=picamv2_jpeg_path,
        camera_version=PiCameraVersion.V2,
        sensor_mode=0
    )

    def test_extracts_bayer_order(self):
        assert self.bayer_order == BayerOrder.BGGR

    def test_extracts_raw_data(self):
        # Spot-check some known pixels
        assert self.bayer_array[0][0] == 91
        assert self.bayer_array[-1][-1] == 65

        # Compare to full np array
        actual = self.bayer_array
        expected = np.load(picamv2_BGGR_bayer_array_path)

        np.testing.assert_array_equal(actual, expected)


class TestBayerArrayTo3D:
    bayer_array = np.array([
        [1, 2],
        [3, 4],
    ])

    @pytest.mark.parametrize('bayer_order,expected', [
        (BayerOrder.RGGB, np.array([
            [[1, 0, 0], [0, 2, 0]],
            [[0, 3, 0], [0, 0, 4]],
        ])),
        (BayerOrder.GBRG, np.array([
            [[0, 1, 0], [0, 0, 2]],
            [[3, 0, 0], [0, 4, 0]],
        ])),
        (BayerOrder.BGGR, np.array([
            [[0, 0, 1], [0, 2, 0]],
            [[0, 3, 0], [4, 0, 0]],
        ])),
        (BayerOrder.GRBG, np.array([
            [[0, 1, 0], [2, 0, 0]],
            [[0, 0, 3], [0, 4, 0]],
        ])),
    ])
    def test_splits_to_rgb_using_bayer_order(self, bayer_order, expected):
        actual = module.bayer_array_to_3d(self.bayer_array, bayer_order)

        np.testing.assert_array_equal(actual, expected)

    def test_integration(self):
        bayer_array = np.load(picamv2_BGGR_bayer_array_path)
        actual = module.bayer_array_to_3d(bayer_array, BayerOrder.BGGR)

        expected = np.load(picamv2_3d_path)

        np.testing.assert_array_equal(actual, expected)


class TestBayerArrayToRGB:
    @pytest.mark.parametrize('bayer_order,expected', [
        (BayerOrder.RGGB, np.array([
            [[1, 2.5, 5]],
        ])),
        (BayerOrder.GBRG, np.array([
            [[3, 3, 2]],
        ])),
        (BayerOrder.BGGR, np.array([
            [[5, 2.5, 1]],
        ])),
        (BayerOrder.GRBG, np.array([
            [[2, 3, 3]],
        ])),
    ])
    def test_splits_to_rgb_using_bayer_order(self, bayer_order, expected):
        bayer_array = np.array([
            [1, 2],
            [3, 5],
        ])
        actual = module.bayer_array_to_rgb(bayer_array, bayer_order)

        np.testing.assert_array_equal(actual, expected)

    def test_uneven_shape_raises(self):
        bayer_array = np.array([
            [1, 2, 3],
            [4, 5, 6],
        ])

        # pytest uses regex to match error messages
        expected_error_message = r'Incoming data is the wrong shape: width \(3\) is not a multiple of 2'
        with pytest.raises(ValueError, match=expected_error_message):
            module.bayer_array_to_rgb(bayer_array, BayerOrder.RGGB)

    def test_integration(self):
        bayer_array = np.load(picamv2_BGGR_bayer_array_path)
        actual = module.bayer_array_to_rgb(bayer_array, BayerOrder.BGGR)

        expected = np.load(picamv2_rgb_path)

        np.testing.assert_array_equal(actual, expected)


class TestUnpack10BitValues:
    def test_unpack_10bit_values(self):
        # In the camera's RAW format, every 5 bytes in the source data contains the high 8-bits of
        # 4 values followed by the low 2-bits of 4 values packed into 5th byte.
        # Spec: https://linuxtv.org/downloads/v4l-dvb-apis-new/uapi/v4l/pixfmt-srggb10p.html
        input_five_byte_set = [
            # Note: In python, "0b" appears to the left of any binary value.
            0b11111111,
            0b10010010,
            0b01001001,
            0b00000000,
            # The fifth byte is made up of the 2-bit sets that will be paired with the first 4 bytes
            # In this test those 2-bit sets will be
            # 0b11 for byte 3, 0b10 for byte 2, 0b01 for byte 1, 0b00 for byte 0
            # Mash those together and you get...
            0b11100100
        ]
        # Those 5 bytes are converted to 10-bit values
        # Each 10-bit output value will be made up of 8 bits from one of the first 4 input bytes
        # and 2 bits from the 5th input byte.
        # The spec linked above determines which bits end up where, but let's also lay it out inline:
        expected_ten_bit_outputs = [
            # The rightmost (least significant) 2-bits of the fifth source byte (0b00 in this case)
            # should end up as the low bits of the first input byte
            0b1111111100,
            # The second input byte gets the next 2 least significant bits (0b01) from the fifth byte
            0b1001001001,
            # The third byte gets the next 2 (0b10)
            0b0100100110,
            # And finally the fourth byte gets the leftmost, most-significant 2 bits (0b11)
            0b0000000011,
        ]

        # In the image array, there are multiple sets of these bytes to unpack, in a 2-dimensional array
        # Let's just do a basic job of simulating that by repeating the same values a few times.
        mock_pixel_bytes_2d = np.array(
            [
                input_five_byte_set * 2,
                input_five_byte_set * 2,
            ],
            dtype=np.uint8,
        )

        expected = np.array(
            [
                expected_ten_bit_outputs * 2,
                expected_ten_bit_outputs * 2,
            ],
            dtype=np.uint16,
        )

        actual = module._unpack_10bit_values(mock_pixel_bytes_2d)

        np.testing.assert_array_equal(actual, expected)

    def test_unpack_10bit_values__correct_shape_doesnt_raise(self):
        mock_pixel_bytes_2d = np.zeros((10, 25)).astype(np.uint8)
        module._unpack_10bit_values(mock_pixel_bytes_2d)

    def test_unpack_10bit_values__incorrect_shape_raises(self):
        # pytest uses regex to match error messages
        expected_error_message = r'Incoming data is the wrong shape: width \(26\) is not a multiple of 5'
        with pytest.raises(ValueError, match=expected_error_message):
            mock_pixel_bytes_2d = np.zeros((10, 26))
            module._unpack_10bit_values(mock_pixel_bytes_2d)


class TestPixelBytesToArray:
    def test_pixel_bytes_to_array(self):
        mock_header = MagicMock(
            height=16,
            width=16,
            padding_right=0,
            padding_down=0,
        )
        # Build up an array of length 512 to make it reshapeable into the default minimum 32x16 padded shape

        # This group of five bytes unpacks to [0b1001, 0b1001, 0b1001, 0b1001]:
        # all the "0b10"s get "01" added as low bits, ending as "0b1001"
        five_byte_group = [0b10, 0b10, 0b10, 0b10, 0b01010101]
        expected_output_byte = 0b1001
        # It takes 4 such sets of 5 bytes to unpack to 16 pixels of data
        # The remaining 12 bytes are padding (to get the minimum width of 32) that will be cropped
        mock_32_byte_row = five_byte_group * 4 + [0] * 12
        mock_1D_pixel_array = np.array(mock_32_byte_row * 16, dtype=np.uint8)

        # After unpacking, I expect all values in the array to be "9"
        expected = np.ones((16, 16), dtype=np.uint16) * expected_output_byte

        actual = module._pixel_bytes_to_array(mock_1D_pixel_array, mock_header)

        np.testing.assert_array_equal(actual, expected)


# Integration test using a known image
class TestPiRawBayer:
    def test_extracts_raw_data(self):
        raw_bayer = module.PiRawBayer(
            filepath=picamv2_jpeg_path,
            camera_version=PiCameraVersion.V2,
            sensor_mode=0
        )

        assert raw_bayer.bayer_order == BayerOrder.BGGR

        # Spot-check some known pixels
        assert raw_bayer.bayer_array[0][0] == 91
        assert raw_bayer.bayer_array[-1][-1] == 65

        # Compare to full np array
        actual = raw_bayer.bayer_array
        expected = np.load(picamv2_BGGR_bayer_array_path)

        np.testing.assert_array_equal(actual, expected)

    def test_rgb_array_property(self, mocker):
        mocker.patch.object(module, 'bayer_array_to_rgb').return_value = sentinel.rgb_array

        raw_bayer = module.PiRawBayer(
            filepath=picamv2_jpeg_path,
            camera_version=PiCameraVersion.V2,
            sensor_mode=0
        )

        assert raw_bayer.to_rgb() == sentinel.rgb_array

    def test_array_3d_property(self, mocker):
        mocker.patch.object(module, 'bayer_array_to_3d').return_value = sentinel.array_3d

        raw_bayer = module.PiRawBayer(
            filepath=picamv2_jpeg_path,
            camera_version=PiCameraVersion.V2,
            sensor_mode=0
        )

        assert raw_bayer.to_3d() == sentinel.array_3d


picamraw/resolution_test.py
from . import resolution as module


class TestPiResolution:
    def test_pad__adds_default_padding(self):
        actual = module.PiResolution(1920, 1080).pad()
        expected = module.PiResolution(width=1920, height=1088)

        assert actual == expected

    def test_pad__adds_custom_padding(self):
        actual = module.PiResolution(100, 200).pad(9, 11)
        expected = module.PiResolution(width=108, height=209)

        assert actual == expected


picamraw/resolution.py
from collections import namedtuple


class PiResolution(namedtuple('PiResolution', ('width', 'height'))):
    '''
    A :func:`~collections.namedtuple` derivative which represents a resolution
    with a `width` and `height`.

    Attrs:
        width: The width of the resolution in pixels
        height: The height of the resolution in pixels
    '''

    __slots__ = ()  # workaround python issue #24931

    def pad(self, pad_width=32, pad_height=16):
        '''
        Returns the resolution padded up to the nearest multiple of *pad_width*
        and *pad_height* which default to 32 and 16 respectively (the camera's
        native block size for most operations).

        Args:
            pad_width: Optional - defaults to 32. The multiple to pad the width to.
            pad_height: Optional - defaults to 16. The multiple to pad the height to.

        Returns:
            A `PiResolution` object with the padding applied
        '''
        return PiResolution(
            width=((self.width + (pad_width - 1)) // pad_width) * pad_width,
            height=((self.height + (pad_height - 1)) // pad_height) * pad_height,
        )

    def __str__(self):
        return '%dx%d' % (self.width, self.height)


picamraw/main.py
import ctypes

import numpy as np

from .constants import PiCameraVersion, BayerOrder, BAYER_ORDER_TO_RGB_CHANNEL_COORDINATES
from .resolution import PiResolution


class PiRawBayer:
    ''' Extracts the raw 10-bit bayer data from a Raspberry Pi camera JPEG+RAW file into a 16-bit numpy array

    Attrs:
        bayer_data: The raw bayer data as a 16-bit 2D numpy array
        bayer_order: A `BayerOrder` enum that indicates the bayer pattern used by the `bayer_array`
    '''
    def __init__(self, filepath, camera_version: PiCameraVersion, sensor_mode=0):
        ''' Initializing a PiRawBayer object results in extracting the raw bayer data from the provided JPEG+RAW file.

        Args:
            filepath: The full path of the JPEG+RAW image to extract raw data from
            camera_version: A `PiCameraVersion` enum representing the camera hardware version used to capture the image
            sensor_mode: Optional - defaults to 0. An integer representing the `sensor_mode` used to capture the image.
                See https://picamera.readthedocs.io/en/release-1.13/fov.html#sensor-modes for more information
                on sensor_modes.
        '''
        bayer_array, bayer_order = extract_raw_from_jpeg(filepath, camera_version, sensor_mode)
        self.bayer_array = bayer_array
        self.bayer_order = bayer_order

    def to_3d(self):
        '''
        Returns: A 16-bit 3D numpy array. This array has the same 2D dimensions as the input `bayer_array`, but pulls
            each pixel out into either the R, G, or B channel in the 3rd dimension. Thus, each "pixel" in the output
            array will be [R, G, B] where 2 of R, G, and B are 0 and the other contains a value from bayer_array.
            It determines whether a given pixel is R, G, or B using the provided `bayer_order`.
        '''
        return bayer_array_to_3d(self.bayer_array, self.bayer_order)

    def to_rgb(self):
        '''
        Returns: A 16-bit 3D numpy array. Every 2x2 containing R, G1, G2, B in the original array is collapsed into a
            single [R, (G1+G2)/2, B] pixel. Thus, this array is 1/4 the size of the input `bayer_array` - both width and
            height are halved.
        '''
        return bayer_array_to_rgb(self.bayer_array, self.bayer_order)


BROADCOM_BAYER_ORDER_TO_ENUM = {
    0: BayerOrder.RGGB,
    1: BayerOrder.GBRG,
    2: BayerOrder.BGGR,
    3: BayerOrder.GRBG,
}

# Byte offsets for header (metadata) and pixel data within the raw bayer data
HEADER_BYTE_OFFSET = 176
PIXEL_BYTE_OFFSET = 32768


class BroadcomRawHeader(ctypes.Structure):
    _fields_ = [
        ('name',          ctypes.c_char * 32),
        ('width',         ctypes.c_uint16),
        ('height',        ctypes.c_uint16),
        ('padding_right', ctypes.c_uint16),
        ('padding_down',  ctypes.c_uint16),
        ('dummy',         ctypes.c_uint32 * 6),
        ('transform',     ctypes.c_uint16),
        ('format',        ctypes.c_uint16),
        ('bayer_order',   ctypes.c_uint8),
        ('bayer_format',  ctypes.c_uint8),
    ]


def extract_raw_from_jpeg(filepath, camera_version, sensor_mode):
    ''' Extracts the raw 10-bit bayer data from a Raspberry Pi camera JPEG+RAW file into a 16-bit numpy array

    Args:
        filepath: The full path of the JPEG+RAW image to extract raw data from
        camera_version: A `PiCameraVersion` enum representing the camera hardware version used to capture the image
        sensor_mode: Optional - defaults to 0. An integer representing the `sensor_mode` used to capture the image.
            See https://picamera.readthedocs.io/en/release-1.13/fov.html#sensor-modes for more information
            on sensor_modes.

    Returns: (bayer_data, bayer_order)
        bayer_data: The raw bayer data as a 16-bit 2D numpy array
        bayer_order: A `BayerOrder` enum that indicates the bayer pattern used by the `bayer_array`
    '''

    with open(filepath, mode='rb') as file:
        jpeg_data_as_bytes = file.read()

    raw_bytes = _get_raw_bayer_bytes(jpeg_data_as_bytes, camera_version, sensor_mode)

    # Extract header (metadata) and pixel data using known byte offsets
    header = BroadcomRawHeader.from_buffer_copy(raw_bytes, HEADER_BYTE_OFFSET)

    # Extract the 1D array of 8-bit (1-byte) values that collectively represent the pixel data
    # Note: pixel data is actually 10-bits per pixel, but is packed into 8-bit values
    pixel_bytes = np.frombuffer(raw_bytes, dtype=np.uint8, offset=PIXEL_BYTE_OFFSET)

    bayer_array = _pixel_bytes_to_array(pixel_bytes, header)
    bayer_order = BROADCOM_BAYER_ORDER_TO_ENUM[header.bayer_order]

    return bayer_array, bayer_order


def _guard_attribute_is_a_multiple_of(attribute_name, attribute_value, multiple):
    if not attribute_value % multiple == 0:
        raise ValueError(
            'Incoming data is the wrong shape: {attribute_name} ({attribute_value}) is not a multiple of {multiple}'
            .format(**locals())
        )


def bayer_array_to_3d(bayer_array, bayer_order: BayerOrder):
    ''' Convert the 2D `bayer_array` to a 3D RGB array, in which each value in the original 2D array is
        moved to one of the three R,G, or B channels.

    Args:
        bayer_array: the 2D bayer array to convert
        bayer_order: A `BayerOrder` enum that indicates the bayer pattern used by the `bayer_array`

    Returns:
        A 3D numpy array. This array has the same 2D dimensions as the input `bayer_array`, but pulls each pixel
        out into either the R, G, or B channel in the 3rd dimension. Thus, each "pixel" in the output array will be
        [R, G, B] where 2 of R, G, and B are 0 and the other contains a value from bayer_array.
        It determines whether a given pixel is R, G, or B using the provided `bayer_order`.

    Example:
        bayer_array_to_3d(
            bayer_array=np.array([
                [1, 2],
                [3, 4],
            ]),
            bayer_order=BayerOrder.RGGB
        )
        >>> np.array([
            [[1, 0, 0], [0, 2, 0]],
            [[0, 3, 0], [0, 0, 4]],
        ])
    '''

    # Prepare an empty 3D array that has the same 2D dimensions as the bayer array
    array_3d = np.zeros(bayer_array.shape + (3,), dtype=bayer_array.dtype)

    ((ry, rx), (gy, gx), (Gy, Gx), (by, bx)) = BAYER_ORDER_TO_RGB_CHANNEL_COORDINATES[bayer_order]

    # Keeps pixels in the same 2D location, but separates into RGB channels based on bayer order
    # Increment by 2: a given color will be in every other column in every other row in the bayer array
    # "Seed" this incrementating using the (x,y) coordinates of that color in the first 2x2 corner of the array
    R_CHANNEL_INDEX, G_CHANNEL_INDEX, B_CHANNEL_INDEX = [0, 1, 2]
    array_3d[ry::2, rx::2, R_CHANNEL_INDEX] = bayer_array[ry::2, rx::2]  # Red
    array_3d[gy::2, gx::2, G_CHANNEL_INDEX] = bayer_array[gy::2, gx::2]  # Green
    array_3d[Gy::2, Gx::2, G_CHANNEL_INDEX] = bayer_array[Gy::2, Gx::2]  # Green
    array_3d[by::2, bx::2, B_CHANNEL_INDEX] = bayer_array[by::2, bx::2]  # Blue

    return array_3d


def bayer_array_to_rgb(bayer_array, bayer_order: BayerOrder):
    ''' Convert the 2D `bayer_array` to a 3D RGB array, in which each value in the original 2D array is
        moved to one of the three R,G, or B channels.

    Args:
        bayer_array: the 2D bayer array to convert
        bayer_order: A `BayerOrder` enum that indicates the bayer pattern used by the `bayer_array`

    Returns:
        A 3D numpy array. Every 2x2 containing R, G1, G2, B in the original array is collapsed into a single
        [R, (G1+G2)/2, B] pixel. Thus, this array is 1/4 the size of the input `bayer_array` - both width and height
        are halved.

    Example:
        bayer_array_to_rgb(
            bayer_array=np.array([
                [1, 2],  # R  G1
                [3, 4],  # G2 B
            ]),
            bayer_order=BayerOrder.RGGB
        )
        >>> np.array([
            [[1, 2.5, 4]],  # R, (G1+G2)/2, B
        ])
    '''
    # Initialize a new array that is the expected shape of 1/2 width and height dimensions
    original_height = bayer_array.shape[0]
    original_width = bayer_array.shape[1]

    _guard_attribute_is_a_multiple_of('width', original_width, 2)
    _guard_attribute_is_a_multiple_of('height', original_height, 2)

    rgb_array = np.zeros((int(original_height/2), int(original_width/2), 3))

    ((ry, rx), (gy, gx), (Gy, Gx), (by, bx)) = BAYER_ORDER_TO_RGB_CHANNEL_COORDINATES[bayer_order]

    # Increment by 2: a given color will be in every other column in every other row in the bayer array
    # "Seed" this incrementating using the (x,y) coordinates of that color in the first 2x2 corner of the array
    R_CHANNEL_INDEX, G_CHANNEL_INDEX, B_CHANNEL_INDEX = [0, 1, 2]
    rgb_array[:, :, R_CHANNEL_INDEX] = bayer_array[ry::2, rx::2]
    rgb_array[:, :, G_CHANNEL_INDEX] = (bayer_array[gy::2, gx::2] + bayer_array[Gy::2, Gx::2])/2
    rgb_array[:, :, B_CHANNEL_INDEX] = bayer_array[by::2, bx::2]

    return rgb_array


def _pixel_bytes_to_array(pixel_bytes, header):
    ''' Convert the 1D array of 8-bit values ("packed" 10-bit values) to a 2D array of 10-bit values. Every 5 bytes
        contains the high 8-bits of 4 values followed by the low 2-bits of 4 values packed into the fifth byte
    '''
    # Reshape and crop the data. The crop's width is multiplied by 5/4 to deal with the packed 10-bit format;
    # the shape's width is calculated in a similar fashion but with padding included
    crop = PiResolution(
        header.width * 5 // 4,
        header.height
    )

    # Honestly I don't entirely grok this code - it seems to imply that the raw data comes with extra (random?)
    # bytes that form padding along the right and bottom edges of the image. To get the actual image data, this padding
    # must be accounted for when reshaping the 1D array of bytes into a 2D array, and then discarded by cropping to the
    # actual image resolution.
    shape = PiResolution(
        (((header.width + header.padding_right) * 5) + 3) // 4,
        (header.height + header.padding_down)
    ).pad()

    pixel_bytes_2d = pixel_bytes.reshape((shape.height, shape.width))[:crop.height, :crop.width]

    array = _unpack_10bit_values(pixel_bytes_2d)

    return array


def _unpack_10bit_values(pixel_bytes_2d):
    ''' Unpack 10-bit values from 8-bit values.
        Every 5 bytes in the input data corresponds to 4 10-bit values in the output.
        The 5 input bytes consist of the high 8 bits of the 4 output values followed by the low 2 bits of
        the output values packed into the fifth byte

        Args:
            pixel_bytes_2d: 2d numpy array where the 2nd dimension arrays are encoded per this spec:
                https://linuxtv.org/downloads/v4l-dvb-apis-new/uapi/v4l/pixfmt-srggb10p.html
        Returns:
            2d numpy array containing 10-bit values unpacked from pixel_bytes_2d, stored with dtype np.uint16.
    '''
    input_height, input_width = pixel_bytes_2d.shape

    # This code assumes that bytes in each row come in sets of 5. If the width is not a multiple of 5, it breaks.
    _guard_attribute_is_a_multiple_of('width', input_width, 5)

    # Set up the output array with the correct shape: same height but width reduced by 4/5
    output_data = np.zeros(
        shape=(
            input_height,
            # Every 5 bytes in the input will be turned into 4 items in the output
            input_width * 4 // 5
        ),
        # This will be filled with 10 bit values, but uint16 is the closest numpy has to offer.
        dtype=np.uint16
    )

    # Method to populate the output array:
    # In each 5-byte set, byte 0 gets the lowest bits from byte 4, byte 1 gets the next lowest bits from byte 4, etc.
    # We can efficiently process all of the "byte 0"'s together, all of the "byte 1"'s together,
    # and so on through the "byte 3"s.
    # For lack of a better term, we'll call these the "byte cohorts" 0 through 3.

    # First, set aside cohort 4: the bytes that will be unpacked into the low bits to go with the first 4 bytes
    cohort_4 = pixel_bytes_2d[:, 4::5]
    for byte_cohort_index in range(4):
        # High bits come from the input array,
        # shifted left by two bits to make room for the low 2-bits which will come from the 5th byte
        high_bits = pixel_bytes_2d[:, byte_cohort_index::5].astype(np.uint16) << 2

        # Now process bits from cohort 4 and unpack the appropriate ones to be our low 2 bits:
        # Shift the bits over so that the relevant ones are in the rightmost (lowest 2 bits) position
        # eg. for byte 1, 0b00001100 -> 0b11
        shifted_bits_from_cohort_4 = cohort_4 >> (byte_cohort_index * 2)
        # Mask the relevant ones (the lowest 2)
        masked_low_bits = shifted_bits_from_cohort_4 & 0b11

        # Finally, put the masked low bits into their place in the output array
        output_data[:, byte_cohort_index::4] = high_bits | masked_low_bits

    return output_data


# Size of the block of the raw bayer data (in bytes) within the full JPEG+RAW file
RAW_BLOCK_SIZE_BY_VERSION_AND_MODE = {
    PiCameraVersion.V1: {
        0: 6404096,
        1: 2717696,
        2: 6404096,
        3: 6404096,
        4: 1625600,
        5: 1233920,
        6: 445440,
        7: 445440,
    },
    PiCameraVersion.V2: {
        0: 10270208,
        1: 2678784,
        2: 10270208,
        3: 10270208,
        4: 2628608,
        5: 1963008,
        6: 1233920,
        7: 445440,
    },
}


def _get_raw_block_size(camera_version, sensor_mode):
    return RAW_BLOCK_SIZE_BY_VERSION_AND_MODE[camera_version][sensor_mode]


def _get_raw_bayer_bytes(jpeg_data_as_bytes, camera_version, sensor_mode):
    ''' Extract the bytes that represent the raw bayer data from the contents of a JPEG+RAW file '''
    # The raw bayer data is at the end of the file, so extract an appropriately-sized block of data from the end
    raw_block_size = _get_raw_block_size(camera_version, sensor_mode)
    raw_bytes = jpeg_data_as_bytes[-raw_block_size:]

    # Bayer data should start with 'BCRM'
    if raw_bytes[:4] != b'BRCM':
        raise ValueError('Unable to locate Bayer data at end of buffer')

    return raw_bytes
